docs(howtos): add AI-ready content guidelines (#8918)
camunda/camunda-docs: howtos/ai-ready-guidelines.md

# AI-ready content guidelines

This guide describes how to write Camunda docs content that's reliably consumable by AI assistants and by RAG/agent pipelines.

It supplements the [Camunda style guide](technical-writing-styleguide.md) and the [contributor documentation guidelines](documentation-guidelines.md).

### 1. Start each page and section with a direct answer

> The first sentence under any heading (including the H1) should directly answer "what is this?" or "what does this section/page do?" in one quotable line.

Reviewer: ```suggestion
The first sentence under any heading (including the H1) should be quotable and directly answer "what is this?" or "what does this section/page do?".
```

I'm curious about how strict we want to be with this.

[Here's an example](http://localhost:3000/docs/next/components/agentic-orchestration/evaluate-agents/monitor-ai-agents/#step-2-open-the-process-instance-in-operate):

```md
## Step 2: Open the process instance in Operate

1. Open [Operate](http://localhost:3000/docs/next/components/operate/operate-introduction/).
2. Locate the process instance created by your prompt. See [view a deployed process](http://localhost:3000/docs/next/components/operate/userguide/basic-operate-navigation/#view-a-deployed-process) for more details.
3. Open your process instance view by clicking on its process instance key.
```

There's no lead-in one liner, but the heading, itself, so clearly states what the section is about. What kind of quotable one liner that directly answers "what is this?" or "what does this section/page do?" would we add to this that doesn't:

- Impact readability (isn't repetitive and awkward).
- Violate the don't repeat yourself and omit needless words principles that we apply in most other cases.

Reviewer 2: Anecdotally, it's always been good practice for search engines etc that will use the first sentence under a heading as the shortdesc in a result, so I've always leaned towards including a sentence underneath a heading.

Reviewer: Ok, thanks for the background! If that's what's best, we can resolve this conversation. 👍

I haven't seen support for that in the past. For example, Google suggests to [create people-first content](https://developers.google.com/search/docs/fundamentals/creating-helpful-content) as the best strategy for SEO and really only mentions the main heading (which we already have guidance for), not subsequent headings. I don't typically see H2s and H3s in search results either. However, I can obviously be wrong about that or missing some data that shows it really does matter.

Reviewer 2: It can be more important in a content reuse model - for example, a section might be surfaced at a higher heading level in another document if you have content used cross-document - although we don't currently have much reuse in docs-as-code/docusaurus, it's still good to future-proof the content in case we move into a CCMS at any point in the future.

- Do not skip levels.
- Within a page, sibling headings at the same depth should describe sibling concepts.

### 5. Put structured data in tables, not prose

Anything that is fundamentally a list of fields, options, parameters, or values goes in a Markdown table.

### 6. Critical information must appear as text, not only in screenshots or diagrams

Agents can't perform optical character recognition. If the only place a binding name, field value, or required setting appears is inside an image, that information is invisible to AI assistants.

- Use images as **supplementary illustration**, never as the only source of a value the reader (human or agent) needs to act on.
- For every diagram, include a short prose description of what it shows.
- For every screenshot of a UI form, include the field names and their values inline in the text before or after the image.

### 7. Never put critical information only inside a tab

Agents typically see only the active tab when a page is rendered, and Markdown-export pipelines often serialize only the default tab. Content inside non-default tabs becomes invisible.

> If prerequisites or steps differ by environment or programming language, list both flat in the page body, clearly labeled. For example:

Reviewer: Is this a reasonable expectation? The [properties reference page](https://docs.camunda.io/docs/next/self-managed/components/orchestration-cluster/core-settings/configuration/properties/), for example, is already huge with the tabs. Flattening this content would lead to a page almost twice the size (three times the size for docs that also include helm configs). And for programming languages, if feels like it would really get in users' way if they had to skip over all the code blocks for programming languages that aren't relevant to their stack.

If the LLMs are consuming the markdown export, rather than the generated page, I suggest we: 

1. Try to figure out a solution to make the content in tabs visible.
2. If that fails, I suggest we come up with a new IA that separates content based on these differences (SM vs SaaS, programming languages, etc). If we do that, we should also come up with a solution that reduces content duplication as that could lead to a maintenance nightmare.

Reviewer: Just to provide some extra context for my concern. As far as I understand, LLMs will mostly be looking at:

- https://docs.camunda.io/llms.txt
- https://docs.camunda.io/llms-full.txt

for ingesting the docs content. In that case, I am confused why the LLM would have trouble seeing the content in tabs.

Author: Adding context since Alex and I already discussed this in another channel.

AI agents tend to read Markdown, but that’s only a preference - and only when they know it’s available. They can still fetch content directly from Camunda Docs as HTML, depending on where and how the agent is invoked. These are the cases where flattening tabs might be relevant.

For example, if a user asks something related to Camunda Docs in the standard ChatGPT or Claude browser UI, the LLM isn’t aware of the site instructions behind the scenes and usually fetches the HTML.

However, after testing the case above, the LLM returned results from the different available tabs.

Since we prefer to use Tabs across our docs for better UX and readability, we’re dropping this suggestion from the guidelines.

Reviewer 2: Can I add in my two cents worth here - critical info should never be hidden in tabs for human users, so I think this might be a case where the human should override the agent UX as it's good practice on the web - humans should never be given the chance to miss critical info via the UI, or made to hunt for something, so for me tabs should only really be used for code samples and that kind of thing ideally.

Author: > critical info should never be hidden in tabs for (human) users

I agree with you, but I think that sort of statement probably fits better in our main style guide than in these guidelines for AI agents, since the former is also consumed by them.

Author: @camunda/tech-writers Another question is whether we need these specific AI-ready guidelines, or whether we want to merge them into our main style guide.

It supplements the [Camunda style guide](technical-writing-styleguide.md) and the [contributor documentation guidelines](documentation-guidelines.md).

### 1. Start each page and section with a direct answer

The first sentence under any heading (including the H1) should directly answer "what is this?" or "what does this section/page do?" in one quotable line.

Aim for one self-contained sentence, ideally under 25 words / ~150 characters. Agents extract the first sentences after a heading; that's what gets cited and indexed.

- **Bad:** A heading "Process instance modification" followed by three paragraphs of history and rationale before defining what process instance modification is.
- **Good:** A heading "Modify a process instance", followed immediately by: "With process instance modification, you can change the active state of a running process instance without redeploying the process."

### 2. Make every section self-contained

A section must make sense in isolation. RAG systems fetch sections, not whole pages, and agents follow links one at a time.

- Don't write "as mentioned above", "see the previous section", "now that you've configured X".

> - Repeat the prerequisite or context inline. Short, slightly redundant prose beats implicit references.

Reviewer 3: I would maybe add some examples here, as it seems to me that "short, slightly redundant" is a bit subjective.

Author: Rephrased and added an example

- If you need to refer to another page, link to it with a descriptive label, not "see here".

### 3. Use descriptive, specific headings

Headings are how agents recover document structure. Vague headings break that recovery.

- **Bad:** "Configuration", "Overview", "Examples", "How it works"
- **Good:** "Configure the Slack outbound connector", "How Zeebe assigns jobs to workers", "Example: Schedule a recurring timer event"

Reuse the same heading wording when the same concept appears in multiple pages so retrieval can cluster them.

### 4. Keep heading hierarchy consistent

> - Use H1 once, for the page title (Docusaurus generates it from frontmatter `title`).

Reviewer 3: Maybe I'm misunderstanding the guideline, but I would rephrase it slightly. Because, right now, we technically do not use an h1 heading, right? So I am not too sure if that particular bullet encourages the same approach or requires the use of an h1 heading added manually.

Author: Good point! Rephrased it

- Do not skip levels.
- Within a page, sibling headings at the same depth should describe sibling concepts.

### 5. Put structured data in tables, not prose

> Anything that is fundamentally a list of fields, options, parameters, or values goes in a Markdown table. Prose descriptions of parameters scatter the contract across paragraphs and make field-level extraction unreliable.

Reviewer: > Prose descriptions of parameters scatter the contract across paragraphs and make field-level extraction unreliable.

I don't really understand what this means. Can you clarify this reasoning?

Author: If parameter information is hidden in prose instead of shown in a structured format, it’s harder to read/tell which details apply to each field.

Dropping the sentence, since it isn’t really needed to understand this recommendation.

### 6. Critical information must appear as text, not only in screenshots or diagrams

> Agents can't perform optical character recognition.. If the only place a binding name, field value, or required setting appears is inside an image, that information is invisible to AI assistants.

Reviewer 4: ```suggestion
Agents can't perform optical character recognition. If the only place a binding name, field value, or required setting appears is inside an image, that information is invisible to AI assistants.
```

Reviewer 4: Otherwise everything looks good!

- Use images as **supplementary illustration**, never as the only source of a value the reader (human or agent) needs to act on.
- For every diagram, include a short prose description of what it shows.
- For every screenshot of a UI form, include the field names and their values inline in the text before or after the image.

### 7. Never put critical information only inside a tab

Agents typically see only the active tab when a page is rendered, and Markdown-export pipelines often serialize only the default tab. Content inside non-default tabs becomes invisible.

> - **Allowed:** Code examples in different languages across tabs (the concept is repeated, only the syntax changes).

Reviewer: Isn't the syntax important for an Agent?

Author: Agree - updated

- **Not allowed:** Prerequisites split across SaaS vs Self-Managed tabs, configuration steps that exist only in one tab, error troubleshooting that only appears under one variant.

If prerequisites or steps differ by environment, list both flat in the page body, clearly labeled. For example:

```markdown
## Prerequisites

**For SaaS:**

- A Camunda 8 SaaS account.
- ...

**For Self-Managed:**

- A running Self-Managed installation.
- ...
```

### 8. Keep code blocks complete and self-contained

- Always close fences (` ``` `).
- Include the language tag (` ```yaml `, ` ```java `, ` ```bash `).

> - Don't split a single example across multiple unrelated fences with prose in between if the prose isn't part of the example.

Reviewer: > ...if the prose isn't part of the example

Can you expand on this?

Author: I have rephrased it. I think it reads better now
